### DISCUSSION TITLE
August Release - Multi-step forms, K8s support, and in-app modals

### DISCUSSION THREAD

**Discussion by joebudi** [Budibase/budibase#2508] [Announcements]
The August release increases the power and flexibility of forms within Budibase. And, I’m pleased to announce, the team has delivered what I believe is a best-in-class form experience. One that’s flexible, beautiful and dynamic! There’s a number of other incredible enhancements, so please read on.

## Multi-step forms + multi-select data type and component + autocomplete

https://user-images.githubusercontent.com/49767913/131519600-5683c019-4019-49ed-8339-5031a1336deb.mov

Budibase now supports multi-step forms and wizard workflows with the new Form Step component and actions. Model complex workflows complete with form step validation with our new flexible form building experience. We've also added a new multi-select data type, a multi-select component, and autocomplete functionality.

## Custom form validation
Users now have total control over form validation. You can set up a whole host of form validation rules - including regex support and some built in validations common with form based apps. Provide custom validation messages for each of your conditions for total validation control and tailored user feedback.

## Dynamic form options
Take total control over your form options. Load form options from API calls, and even hard code your own.

## New and improved MongoDB connector (Community Contribution)
The MongoDB connector has been improved to support all the high level API methods of the node mongoDB connector. Operators such as find, findOne, count, distinct, aggregate, insertMany, and lots more, are now supported. Thanks to faroutchris for this [PR](https://github.com/Budibase/budibase/pull/1963)  🥳
 
## Kubernetes support
Budibase now supports a Kubernetes based deployment. You can use our helm chart to deploy Budibase directly into your kubernetes cluster.
[Link to K8s docs](https://docs.budibase.com/self-hosting/hosting-methods/kubernetes)

## Open screens in modal

<img width="1680" alt="Screenshot 2021-08-30 at 15 29 43" src="https://user-images.githubusercontent.com/49767913/131519870-a23f2649-4780-4b7c-bdd1-00c6ff9617db.png">

Previous to this feature, you had to open a new screen to edit a record. This was a tedious process for quick analysis and making small changes. With this new feature, you can now open a record in a modal in your application. This means you do not have to route to a completely separate page when navigating to another screen. Powerful for editing workflows and much more. Implementing this new feature in your apps is as simple as clicking a checkbox.

##  Bug fixes + QoL Improvements
Add sizes to boolean field and checkbox: https://github.com/Budibase/budibase/pull/2239
Add headers to webhook automations: https://github.com/Budibase/budibase/pull/2244
Rename an app: https://github.com/Budibase/budibase/pull/2135

## Coming up next
- Cloud Sandbox - try Budibase in under 30 seconds in a managed environment for free

**Comment by KernOil** [comment]
Great job guy love the new features! 
Custom Validation is such a huge upgrade Great work! 
Really looking forwards to playing around with Open screens in modal!

**Comment by MarkOSullivan94** [comment]
Excellent additions! Really looking forward to trying out the multi-step forms 😍 

What version will this all be available in so I know to keep an eye out for it?

**Reply by joebudi** [reply]
v0.9.117

**Comment by KernOil** [comment]
Is there a way to define the size of the "Open screens in modal" window?

**Reply by joebudi** [reply]
Hey!

This is not possible right now. Would you like to increase/decrease the height/width?

**Reply by KernOil** [reply]
yes, I think it would be good to be able to fit the entire form in without scrolling.

 secondly, I can not figure out how to navigate to a new page when submitting a form in the pop-up window. I can get it to navigate to the new page in the window or simply close the window I can not get it to save, close the window, and navigate to a new page. is there a trick to it?

**Reply by joebudi** [reply]
That's a good point. I think increasing the height is probably a good option to have.

I'm not sure this is possible. If you have tried chaining actions; close modal, navigate to, and it has still not worked then I expect this is not possible with actions. Is this right @aptkingston ?

**Reply by aptkingston** [reply]
Yea this isn't possible - closing the modal will effectively stop all further button actions, because they were being executed inside the modal (it's an iframe instance). I'll have a look and see if there's another way to handle this!

**Reply by sergey-dme** [reply]
Hi @aptkingston, Currently, if a user has custom headers/footers in his layout -- they displaying/appear in modal screens. It would be great if the only **Screen Slot** part of the layout will be displaying in modal screens unless it switched to full screen. Thank you for consideration

**Reply by joebudi** [reply]
Hmm. This seems strange as the screenshot should be the only part that shows. In your use cases, are you using an empty layout?

**Reply by aptkingston** [reply]
@sergey-dme is correct that all other components in the layout are visible - the only thing that is hidden is the built in navigation bar. This was for 2 reasons:
1. The layout is important for laying out the app correctly (but this could maybe be replaced by a different wrapper component if we only wanted to show the screen slot)
2. Users might want to have things in their layout that they don't want to always be hidden inside a modal. This is harder to solve, because it's not really feasible to add a setting to every component saying "hide in screen peek".

One solution I can propose is adding a context binding to represent if the app is currently inside a modal screen or not. Then you could use conditional UI to hide your header (or any other component you want) when inside the modal only. Would that be acceptable Sergey?

**Reply by sergey-dme** [reply]
@aptkingston Absolutely! The proposed solution with bind for modal to use it with Conditional UI would be acceptable and work perfectly! Thank you Andrew

**Reply by sergey-dme** [reply]
> Hmm. This seems strange as the screenshot should be the only part that shows. In your use cases, are you using an empty layout?

Hi @joebudi Yep, I checked this new amazing feature using the empty layout too for my modal screen. Using the empty one was problematic, because if a screen switches to full, no headers/navigation are available. 
Along with the bind for modal and size options, I hope in future it will be also possible to change the background of the modal screen. I tried to use CSS `--spectrum-dialog-confirm-background-color` and --`spectrum-alias-background-color-default` to change the bg, it worked, but affected also on confirmation window as the same CSS class is using. Hope this option will be available in Theming but... user may want to use different bg for different modal screens.. even the transparent one

**Reply by KernOil** [reply]
@aptkingston I think your contextual binding is a great idea.

**Comment by erakastrati** [comment]
Great job guys! I was looking forward for these new additions.
Keep rocking 👏🏼

**Reply by joebudi** [reply]
Thank you @erakastrati !

**Comment by Fabian1410** [comment]
Cant update :( 
![image](https://user-images.githubusercontent.com/65181345/132088172-5aed6443-3975-4af0-be73-59e7b4ae1a0a.png)

**Reply by Fabian1410** [reply]
Also CLI update doenst work 

`? Do you wish to update you docker-compose.yaml and envoy.yaml? Yes
Error - Failed to start - logs written to file: docker-error.log
`

**Reply by shogunpurple** [reply]
What does the docker-error.log say?

**Reply by Fabian1410** [reply]
services.watchtower-service.ports is invalid: Port ranges don't match in length

**Reply by shogunpurple** [reply]
Ah yes, you will need to add 

`WATCHTOWER_PORT=6161 `

to your `hosting.properties` file. 

Then run again

**Reply by Fabian1410** [reply]
I dont have a hosting.properties file, and no .env folder

**Reply by shogunpurple** [reply]
.env is a file, not a folder. 

Do you not have either? Could you show me the contents of your Budibase directory?

**Reply by Fabian1410** [reply]
I fixed it :)

**Comment by amigoni** [comment]
Is this version released yet? I can't seem to find how to check my version from the UI and when I press update it says I'm up to date. Thanks

**Reply by shogunpurple** [reply]
Hi @amigoni - this version has been released. 
When you go to the screen with the update button, is your current Budibase version not shown?

If you aren’t on the latest and it says it’s up to date, it sounds like you might have a problem with your watchtower (update) service.

**Reply by Fabian1410** [reply]
@amigoni check my question before :)

**Reply by amigoni** [reply]
Correct, I don't see any version. I create an instance on Digital Ocean. Sorry not much of a devops person. I see the question but I don't know where to find CLI commands to update. Any pointers?

**Reply by shogunpurple** [reply]
Hey @amigoni - if you SSH into your digitalocean droplet, you can run the update commands from there:

`budi hosting --update`

**Reply by lars0407** [reply]
The command doesnt work...

**Reply by shogunpurple** [reply]
@lars0407 ok, what happened when you ran the command? Any output in the terminal?

**Reply by lars0407** [reply]
It needs to be --update or?

**Reply by shogunpurple** [reply]
Yep, it's `budi hosting --update`

**Reply by lars0407** [reply]
However the update fails to start

**Reply by lars0407** [reply]
![mintty_hi7jEZFA8Z](https://user-images.githubusercontent.com/74684066/132844915-f5268770-7c07-42f5-b33e-8c9676743910.png)

**Reply by shogunpurple** [reply]
`cat docker-error.log`

**Reply by lars0407** [reply]
![mintty_Z1YwBFnTKV](https://user-images.githubusercontent.com/74684066/132845129-edbbda7a-ffd1-4562-b293-8b9cbd68704a.png)

**Reply by lars0407** [reply]
its hosted on digitalocean with your onclick setup

**Reply by shogunpurple** [reply]
`ls`

**Reply by shogunpurple** [reply]
then `cat hosting.properties`

**Reply by lars0407** [reply]
No such file or directory

**Reply by shogunpurple** [reply]
`echo WATCHTOWER_PORT=6161 >> .env`
`budi-hosting --update`

**Reply by lars0407** [reply]
![mintty_HlXjs9E1ha](https://user-images.githubusercontent.com/74684066/132845796-1e8197a7-4e9c-4fb8-a0cb-b714303affb7.png)

**Reply by lars0407** [reply]
same issue

**Reply by shogunpurple** [reply]
can you run a `ls -la` just so I can see what is in your directory? That should have worked. Also the contents of `docker-error.log` again just to see if it's the same issue.

**Reply by lars0407** [reply]
![mintty_SPyFe63350](https://user-images.githubusercontent.com/74684066/132846139-6b6d702e-c85c-4f17-8784-bf880715e912.png)

**Reply by lars0407** [reply]
copy and paste seems to work

**Reply by lars0407** [reply]
sec

**Reply by lars0407** [reply]
ok, worked.. Thanks so much for the help and time

**Reply by shogunpurple** [reply]
No problem - enjoy using budibase!

**Reply by lars0407** [reply]
I will :D

**Reply by amigoni** [reply]
I still don't know how to fix this lost you at "Copy and paste seems to work" copy and paste what?

**Reply by shogunpurple** [reply]
Hi @amigoni - please try the following in your DO droplet.

```
echo "WATCHTOWER_PORT=6161" >> .env
budi hosting --update
```

**Reply by amigoni** [reply]
Gotcha. I will still having issues. Finally figured it out. I know it sounds trivial, but you have to STOP budibase then UPDATE then START. The update command does not stop budibase automatically.
